elastic/apm-server
# gomaxprocs: Use a ticker (30s) to set gomaxprocs

## Motivation/summary

Adds a new goroutine which tries to refresh the gomaxprocs setting every
`30s`. This accounts for cases where the CFS quotas may be refreshed
without the APM Server process being restarted (like in ESS / ECE).

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

## How to test these changes

0. `docker-compose up -d`
1. `cd systemtest/cmd/runapm && go run main.go`, copy the image that is used.
2. Inspect the created container and obtain the `FLEET_ENROLLMENT_TOKEN` from the runapm container (you can stop the runapm container afterwards).
3. Create a `docker-compose.override.yml` in the root of the APM Server with these contents:
```
services:
  apm-server:
    image: USE_BUILDAPM_RESULTING_IMAGE
    cpus: 2.7
    ports:
      - 8200:8200
    environment:
      - "FLEET_URL=https://fleet-server:8220"
      - "FLEET_CA=/etc/pki/tls/certs/fleet-ca.pem"
      - "FLEET_ENROLL=1"
      - "FLEET_ENROLLMENT_TOKEN=USE_RUNAPM_ENROLLMENT_TOKEN"
    healthcheck:
      test: ["CMD-SHELL", "curl -s http://localhost:8200 | grep -q 'true'"]
      retries: 300
      interval: 1s
    depends_on:
      elasticsearch: { condition: service_healthy }
      kibana: { condition: service_healthy }
      fleet-server: { condition: service_healthy }
    volumes:
      - "./testing/docker/fleet-server/certificate.pem:/etc/pki/tls/certs/fleet-ca.pem"
```
4. `docker-compose up -d`. Verify that the apm-server container is reachable.
5. Run: `docker update -cpus=1 <CONTAINER ID>`
6. Observe a log message (In the observability logs UI) indicating that the CPU quota has been updated to 1.

## Related issues

Part of #7967
